core/src/main/scala/sttp/tapir/server/interceptor/metrics/MetricsInterceptor.scala

```diff
@@ -0,0 +1,71 @@
+package sttp.tapir.server.interceptor.metrics
+
+import sttp.monad.MonadError
+import sttp.monad.syntax._
+import sttp.tapir.metrics.Metric
+import sttp.tapir.model.{ServerRequest, ServerResponse}
+import sttp.tapir.server.interceptor.{EndpointInterceptor, ValuedEndpointOutput}
+import sttp.tapir.{DecodeResult, Endpoint, EndpointInput}
+
+class MetricsInterceptor[F[_], B](metrics: List[Metric[F, _]], ignoreEndpoints: Seq[Endpoint[_, _, _, _]])(implicit monad: MonadError[F])
+    extends EndpointInterceptor[F, B] {
+
+  override def onDecodeSuccess[I](
+      request: ServerRequest,
+      endpoint: Endpoint[I, _, _, _],
+      i: I,
+      next: Option[ValuedEndpointOutput[_]] => F[ServerResponse[B]]
+  )(implicit monad: MonadError[F]): F[ServerResponse[B]] = {
+    if (ignoreEndpoints.contains(endpoint)) next(None)
+    else {
+      for {
+        _ <- collectMetrics { case Metric(m, Some(onRequest), _, _) => onRequest(request, m) }
+        response <- next(None)
+        _ <- collectMetrics { case Metric(m, _, Some(onResponse), _) => onResponse(request, response, m) }
+      } yield response
```

Comment:
we should also probably somehow notify the metrics if the effect fails, that is, there's an exception that's being thrown?

that also depends, if the metrics interceptor should wrap the exception-handling one, or the other way round

Reply:
Maybe `MetricsInterceptor` should be always called after `ExceptionInterceptor`?

Reply:
yeah, well ... unless there's an exception during metrics collection ;-) but maybe that should be handled in the server.

If metrics should be called first/last (around the exc interceptor), they have to be added to the front of the interceptors list

Reply:
Another problem: the response is only really "done" (well not even then, but it's more "done" than currently) when the whole response is sent. In terms of akka-http/http4s: when the response stream is completed. I think we won't run away from needing some way of interpreter-specific integration, which would allow us to "plug into" the response stream, getting a callback once it is complete.

In other words - a `ServerResponse` can be created quickly, but with a stream that takes a long time to evaluate